# <center><font color="green">FINAL ASSIGNMENT</font></center>

## PART 0. DOWNLOADING AND PREPARING THE DATA

We import the necessary libraries

In [4]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

Download the data and read it into a pandas dataframe

In [5]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Check for missing values

In [6]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [7]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

Our dataset looks clean. Now we can start to build our model.

# PART 1

We first split data into predictors and target. Then we divide into training and test set.

In [8]:
# split into predictors and target
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

 # number of predictors, we're gonna need it later
n_cols = predictors.shape[1]

# train test split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)



We define the model.

In [10]:
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Build and fit the model.

In [11]:
# build
model = regression_model()
# fit
m = model.fit(X_train, y_train,epochs=50, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 2351.2817
Epoch 2/50
23/23 - 0s - loss: 1362.4167
Epoch 3/50
23/23 - 0s - loss: 1201.5991
Epoch 4/50
23/23 - 0s - loss: 1091.7037
Epoch 5/50
23/23 - 0s - loss: 995.1987
Epoch 6/50
23/23 - 0s - loss: 908.0590
Epoch 7/50
23/23 - 0s - loss: 835.3489
Epoch 8/50
23/23 - 0s - loss: 766.6902
Epoch 9/50
23/23 - 0s - loss: 713.2257
Epoch 10/50
23/23 - 0s - loss: 663.0255
Epoch 11/50
23/23 - 0s - loss: 618.0795
Epoch 12/50
23/23 - 0s - loss: 580.5922
Epoch 13/50
23/23 - 0s - loss: 548.3224
Epoch 14/50
23/23 - 0s - loss: 521.1786
Epoch 15/50
23/23 - 0s - loss: 501.3986
Epoch 16/50
23/23 - 0s - loss: 468.2598
Epoch 17/50
23/23 - 0s - loss: 449.5962
Epoch 18/50
23/23 - 0s - loss: 431.2016
Epoch 19/50
23/23 - 0s - loss: 416.3965
Epoch 20/50
23/23 - 0s - loss: 400.1253
Epoch 21/50
23/23 - 0s - loss: 386.8980
Epoch 22/50
23/23 - 0s - loss: 373.5831
Epoch 23/50
23/23 - 0s - loss: 363.0927
Epoch 24/50
23/23 - 0s - loss: 353.8918
Epoch 25/50
23/23 - 0s - loss: 340.1528
Epoch

We calculate the mean square error.

In [12]:
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)

In [14]:
print('The mean of the mean squared errores is %s' %"{:.2f}".format(mse))

The mean of the mean squared errores is 194.18


Repeat the process 50 times.

In [15]:
mseList = []

for i in range(1,51):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)
    model = regression_model()
    m = model.fit(X_train, y_train,epochs=50, verbose=2)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mseList.append(mse)


aux = np.array(mseList)


mseMeanPart1 = aux.mean()
stdMeanPart1 = aux.std()


Epoch 1/50
23/23 - 1s - loss: 1604.0098
Epoch 2/50
23/23 - 0s - loss: 1295.2705
Epoch 3/50
23/23 - 0s - loss: 1081.6670
Epoch 4/50
23/23 - 0s - loss: 912.5745
Epoch 5/50
23/23 - 0s - loss: 787.3088
Epoch 6/50
23/23 - 0s - loss: 690.3854
Epoch 7/50
23/23 - 0s - loss: 605.7183
Epoch 8/50
23/23 - 0s - loss: 529.4744
Epoch 9/50
23/23 - 0s - loss: 467.2942
Epoch 10/50
23/23 - 0s - loss: 417.4436
Epoch 11/50
23/23 - 0s - loss: 376.8832
Epoch 12/50
23/23 - 0s - loss: 340.0207
Epoch 13/50
23/23 - 0s - loss: 306.4856
Epoch 14/50
23/23 - 0s - loss: 278.3086
Epoch 15/50
23/23 - 0s - loss: 254.6075
Epoch 16/50
23/23 - 0s - loss: 234.2063
Epoch 17/50
23/23 - 0s - loss: 218.0121
Epoch 18/50
23/23 - 0s - loss: 204.8019
Epoch 19/50
23/23 - 0s - loss: 191.0114
Epoch 20/50
23/23 - 0s - loss: 179.9048
Epoch 21/50
23/23 - 0s - loss: 172.7093
Epoch 22/50
23/23 - 0s - loss: 165.7384
Epoch 23/50
23/23 - 0s - loss: 159.6543
Epoch 24/50
23/23 - 0s - loss: 154.5301
Epoch 25/50
23/23 - 0s - loss: 151.5787
Epoch 

In [18]:
print('The mean of the mean squared errores is %s' %"{:.2f}".format(mseMeanPart1))
print('The standard deviation of the mean squared errores is %s' %"{:.2f}".format(stdMeanPart1))


The mean of the mean squared errores is 363.78
The standard deviation of the mean squared errores is 479.68


# PART 2

Now, we normalize the data.

In [37]:
mseListNormalizedData = []

for i in range(1,51):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)
    # normalize
    X_train_norm = (X_train - X_train.mean() / X_train.std())
    X_test_norm = (X_test - X_test.mean() / X_test.std())
    # build, fit and error
    model = regression_model()
    m = model.fit(X_train_norm, y_train,epochs=50, verbose=2)
    y_pred = model.predict(X_test_norm)
    mse = mean_squared_error(y_test, y_pred)
    mseListNormalizedData.append(mse)


aux2 = np.array(mseListNormalizedData)

mseMeanPart2 = aux2.mean()
stdMeanPart2 = aux2.std()


Epoch 1/50
23/23 - 0s - loss: 155677.5156
Epoch 2/50
23/23 - 0s - loss: 56525.1484
Epoch 3/50
23/23 - 0s - loss: 15906.7285
Epoch 4/50
23/23 - 0s - loss: 4119.0879
Epoch 5/50
23/23 - 0s - loss: 1903.6750
Epoch 6/50
23/23 - 0s - loss: 1641.7821
Epoch 7/50
23/23 - 0s - loss: 1569.7848
Epoch 8/50
23/23 - 0s - loss: 1507.9102
Epoch 9/50
23/23 - 0s - loss: 1445.4403
Epoch 10/50
23/23 - 0s - loss: 1380.8993
Epoch 11/50
23/23 - 0s - loss: 1314.9745
Epoch 12/50
23/23 - 0s - loss: 1251.3654
Epoch 13/50
23/23 - 0s - loss: 1186.4009
Epoch 14/50
23/23 - 0s - loss: 1122.2158
Epoch 15/50
23/23 - 0s - loss: 1059.0803
Epoch 16/50
23/23 - 0s - loss: 997.9668
Epoch 17/50
23/23 - 0s - loss: 936.7247
Epoch 18/50
23/23 - 0s - loss: 877.0645
Epoch 19/50
23/23 - 0s - loss: 810.7206
Epoch 20/50
23/23 - 0s - loss: 746.6807
Epoch 21/50
23/23 - 0s - loss: 693.8531
Epoch 22/50
23/23 - 0s - loss: 643.7090
Epoch 23/50
23/23 - 0s - loss: 604.8557
Epoch 24/50
23/23 - 0s - loss: 565.6974
Epoch 25/50
23/23 - 0s - loss:

In [38]:
print('The mean of the mean squared errores is %s' %"{:.2f}".format(mseMeanPart2))
print('The standard deviation of the mean squared errores is %s' %"{:.2f}".format(stdMeanPart2))

The mean of the mean squared errores is 226.01
The standard deviation of the mean squared errores is 169.71


We see that the error and the standard deviation decreased considerably compared to part one

# PART 3

We increment the number of epochs.

In [22]:
listPart3 = []
epochs = 100

for i in range(1,51):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)
    X_train_norm = (X_train - X_train.mean() / X_train.std())
    X_test_norm = (X_test - X_test.mean() / X_test.std())
    model = regression_model()
    m = model.fit(X_train_norm, y_train,epochs=epochs, verbose=2)
    y_pred = model.predict(X_test_norm)
    mse = mean_squared_error(y_test, y_pred)
    listPart3.append(mse)


aux3 = np.array(listPart3)

mseMeanPart3 = aux3.mean()
stdMeanPart3 = aux3.std()




Epoch 1/100
23/23 - 1s - loss: 28271.8438
Epoch 2/100
23/23 - 0s - loss: 3603.9858
Epoch 3/100
23/23 - 0s - loss: 1190.7430
Epoch 4/100
23/23 - 0s - loss: 1113.3674
Epoch 5/100
23/23 - 0s - loss: 1028.6343
Epoch 6/100
23/23 - 0s - loss: 958.6007
Epoch 7/100
23/23 - 0s - loss: 892.2621
Epoch 8/100
23/23 - 0s - loss: 828.2245
Epoch 9/100
23/23 - 0s - loss: 766.2822
Epoch 10/100
23/23 - 0s - loss: 708.7405
Epoch 11/100
23/23 - 0s - loss: 655.4681
Epoch 12/100
23/23 - 0s - loss: 605.0386
Epoch 13/100
23/23 - 0s - loss: 560.1409
Epoch 14/100
23/23 - 0s - loss: 518.6736
Epoch 15/100
23/23 - 0s - loss: 479.2932
Epoch 16/100
23/23 - 0s - loss: 444.7768
Epoch 17/100
23/23 - 0s - loss: 413.0578
Epoch 18/100
23/23 - 0s - loss: 383.0207
Epoch 19/100
23/23 - 0s - loss: 357.3903
Epoch 20/100
23/23 - 0s - loss: 334.2345
Epoch 21/100
23/23 - 0s - loss: 311.4045
Epoch 22/100
23/23 - 0s - loss: 293.9377
Epoch 23/100
23/23 - 0s - loss: 277.1397
Epoch 24/100
23/23 - 0s - loss: 261.1604
Epoch 25/100
23/23 

In [24]:
print('The mean of the mean squared errores is %s' %"{:.2f}".format(mseMeanPart3))
print('The standard deviation of the mean squared errores is %s' %"{:.2f}".format(stdMeanPart3))

The mean of the mean squared errores is 155.98
The standard deviation of the mean squared errores is 107.22


We see that the error and the standard deviation decreased considerably compared to part two

# PART 4

We add hidden layers.

In [25]:
def regression_model2():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))       
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [29]:
listPart4 = []
epochs = 50

for i in range(1,51):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)
    # normalize
    X_train_norm = (X_train - X_train.mean() / X_train.std())
    X_test_norm = (X_test - X_test.mean() / X_test.std())
    # construimos el modelo
    model = regression_model2()
    # ajustamos el modelo
    m = model.fit(X_train_norm, y_train,epochs=epochs, verbose=2)
    y_pred = model.predict(X_test_norm)
    mse = mean_squared_error(y_test, y_pred)
    listPart4.append(mse)


aux4 = np.array(listPart4)

mseMeanPart4 = aux4.mean()
stdMeanPart4 = aux4.std()

Epoch 1/50
23/23 - 1s - loss: 31370.5430
Epoch 2/50
23/23 - 0s - loss: 5150.8286
Epoch 3/50
23/23 - 0s - loss: 1691.4707
Epoch 4/50
23/23 - 0s - loss: 1238.1150
Epoch 5/50
23/23 - 0s - loss: 949.4794
Epoch 6/50
23/23 - 0s - loss: 747.2892
Epoch 7/50
23/23 - 0s - loss: 591.1071
Epoch 8/50
23/23 - 0s - loss: 474.3994
Epoch 9/50
23/23 - 0s - loss: 388.7936
Epoch 10/50
23/23 - 0s - loss: 325.6750
Epoch 11/50
23/23 - 0s - loss: 281.9538
Epoch 12/50
23/23 - 0s - loss: 247.5544
Epoch 13/50
23/23 - 0s - loss: 223.8021
Epoch 14/50
23/23 - 0s - loss: 207.9123
Epoch 15/50
23/23 - 0s - loss: 195.3395
Epoch 16/50
23/23 - 0s - loss: 187.9705
Epoch 17/50
23/23 - 0s - loss: 180.5643
Epoch 18/50
23/23 - 0s - loss: 175.2214
Epoch 19/50
23/23 - 0s - loss: 170.5080
Epoch 20/50
23/23 - 0s - loss: 166.5547
Epoch 21/50
23/23 - 0s - loss: 163.3860
Epoch 22/50
23/23 - 0s - loss: 159.5196
Epoch 23/50
23/23 - 0s - loss: 156.5936
Epoch 24/50
23/23 - 0s - loss: 155.1556
Epoch 25/50
23/23 - 0s - loss: 153.7632
Epoc

In [30]:
print('The mean of the mean squared errores is %s' %"{:.2f}".format(mseMeanPart4))
print('The standard deviation of the mean squared errores is %s' %"{:.2f}".format(stdMeanPart4))

The mean of the mean squared errores is 116.26
The standard deviation of the mean squared errores is 44.22


We see that the error and the standard deviation decreased considerably compared to part two

## RESULTS

Model | Mean | Standard Deviation
:--------: | -------: | -------
Initial | 363.78 | 479.68
Normalized | 226.01 | 169.71
More epochs | 155.98 | 107.22
More hidden layers | 116.26 | 44.22

## DISCUSION

We can see that normalizing the data, increasing the number of epochs and increasing the number of hidden layers can lead to a significant improvement in the performance of the model.